src/transport/manager/mod.rs

```diff
@@ -3266,4 +3369,163 @@ mod tests {
             state => panic!("invalid peer state: {state:?}"),
         }
     }
+
+    #[tokio::test]
+    async fn manager_limits_incoming_connections() {
+        let _ = tracing_subscriber::fmt()
+            .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+            .try_init();
+
+        let (mut manager, _handle) = TransportManager::new(
+            Keypair::generate(),
+            HashSet::new(),
+            BandwidthSink::new(),
+            8usize,
+            ConnectionLimitsConfig::default()
+                .max_incoming_connections(Some(3))
+                .max_outgoing_connections(Some(2)),
+        );
+        // The connection limit is agnostic of the underlying transports.
+        manager.register_transport(SupportedTransport::Tcp, Box::new(DummyTransport::new()));
+
+        let peer = PeerId::random();
+        let second_peer = PeerId::random();
+
+        // Setup addresses.
+        let (first_addr, first_connection_id) = setup_dial_addr(peer, 0);
+        let (second_addr, second_connection_id) = setup_dial_addr(second_peer, 1);
+        let (_, third_connection_id) = setup_dial_addr(peer, 2);
+        let (_, remote_connection_id) = setup_dial_addr(peer, 3);
+
+        // Peer established the first inbound connection.
+        let result = manager
+            .on_connection_established(
+                peer,
+                &Endpoint::listener(first_addr.clone(), first_connection_id),
+            )
+            .unwrap();
+        assert_eq!(result, ConnectionEstablishedResult::Accept);
+
+        // The peer is allowed to dial us a second time.
+        let result = manager
+            .on_connection_established(
+                peer,
+                &Endpoint::listener(first_addr.clone(), second_connection_id),
```

Comment:
With the current code logic in the transport manager, we can allow a maximum of 2 connections.
We can definitely have 2 inbound connections: 
- I believe this is an artifact of each transport negotiating and accepting connections. We may end-up with two consecutive `Endpoint::Listener`. After we handle https://github.com/paritytech/litep2p/issues/186, we can then check if we have an established inbound connection already, since we are most probably going to change the Peerstate a bit 
- If we manage to dial the peer before a second connection happens, the connection is still accepted. These PRs should fix it;
  - https://github.com/paritytech/litep2p/pull/176
  - https://github.com/paritytech/litep2p/pull/179

Reply:
Was the initial design about supporting a maximum of 2 connections, only if the inbound connection happens at the same time with the outbound connection? If so we can prob have a look into the peerstate and modify it :D

Reply:
Have created https://github.com/paritytech/litep2p/issues/189 as followup from this, offhand it seems beneficial to reject a second inbound connection